syntax = "proto2";
package tdrec.protos;

import "tdrec/protos/backbone.proto";
import "tdrec/protos/loss.proto";
import "tdrec/protos/metric.proto";
import "tdrec/protos/module.proto";

message ModelConfig {
    repeated FeatureGroupConfig feature_groups = 2;
    oneof model {
        RankModel rank_model = 11;
        MultiTaskRankModel multi_task_rank_model = 12;
    }
    required BackboneConfig backbone = 3;
    repeated LossConfig losses = 4;
    repeated MetricConfig metrics = 5;
}

message FeatureGroupConfig {
    required string group_name = 1;
    repeated string feature_names = 2;
    optional FeatureGroupType group_type = 3 [default = Deep];
}

enum FeatureGroupType {
    Deep = 1;
    Deep_3D = 2;
    Sequence = 3;
    Sequence_Attention = 4;
}

message RankModel {
    required MLP top_mlp = 1;
}

message MultiTaskRankModel {
    repeated TaskTower task_towers = 1;
    oneof multi_task_model {
        MMoE mmoe = 11;
        DBMTL dbmtl = 12;
        ESMM esmm = 13;
    }
}

message TaskTower {
    // label for the task
    required string label_name = 1;
    // label for the task
    required uint32 tower_input_dim = 2;
    // task specific mlp
    required MLP mlp = 3;
    // training loss weights
    optional float weight = 4 [default = 1.0];
}
